## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
print("Column Names:")
print(df.columns)

Column Names:
Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [3]:
# Use the describe function to gather some basic statistics
print("\nBasic Statistics:")
print(df.describe())



Basic Statistics:
          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.

In [6]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info



<bound method DataFrame.info of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  

In [7]:
# What three item categories had the most entries?
print("\nTop 3 Categories by Entries:")
top_categories = df['category'].value_counts().head(3)
print(top_categories)


Top 3 Categories by Entries:
category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [9]:
# For the category with the most entries, which subcategory had the most entries?
most_common_category = top_categories.index[0]
top_subcategory = df[df['category'] == most_common_category]['subcategory'].value_counts().idxmax()
print(f"\nTop Subcategory in '{most_common_category}':")
print(top_subcategory)



Top Subcategory in 'consumables':
bathroom supplies


In [10]:
# Which five clients had the most entries in the data?
print("\nTop 5 Clients by Entries:")
top_clients = df['client_id'].value_counts().head(5)
print(top_clients)



Top 5 Clients by Entries:
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [11]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = top_clients.index.tolist()
print(f"\nTop 5 Client IDs: {top_client_ids}")


Top 5 Client IDs: [33615, 66037, 46820, 38378, 24741]


In [12]:
# How many total units (the qty column) did the client with the most entries order order?
client_with_most_entries = top_clients.idxmax()
total_units_most_entries = df[df['client_id'] == client_with_most_entries]['qty'].sum()
print(f"\nTotal Units Ordered by Client {client_with_most_entries}: {total_units_most_entries}")


Total Units Ordered by Client 33615: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [14]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
print("Line Subtotal (first 5 rows):")
print(df[['unit_price', 'qty', 'line_subtotal']].head())

Line Subtotal (first 5 rows):
   unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00


In [15]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(lambda x: 7*x if x > 50 else 10*x)
print("\nShipping Price (first 5 rows):")
print(df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head())



Shipping Price (first 5 rows):
   unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20


In [16]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['line_price'] = (df['line_subtotal'] + df['shipping_price']) * 1.0925
print("\nLine Price (first 5 rows):")
print(df[['line_subtotal', 'shipping_price', 'line_price']].head())



Line Price (first 5 rows):
   line_subtotal  shipping_price     line_price
0      115164.00         5512.50  131839.076250
1         523.95          312.90     914.258625
2         527.28          458.64    1077.117600
3        1056.18          356.70    1543.571400
4        3902.00         6500.20   11364.403500


In [17]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
print("\nLine Cost (first 5 rows):")
print(df[['unit_cost', 'qty', 'shipping_price', 'line_cost']].head())



Line Cost (first 5 rows):
   unit_cost  qty  shipping_price  line_cost
0     762.71  105         5512.50   85597.05
1      15.09   21          312.90     629.79
2       7.86   39          458.64     765.18
3      24.85   29          356.70    1077.35
4     108.17   20         6500.20    8663.60


In [18]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['line_price'] - df['line_cost']
print("\nLine Profit (first 5 rows):")
print(df[['line_price', 'line_cost', 'line_profit']].head())



Line Profit (first 5 rows):
      line_price  line_cost   line_profit
0  131839.076250   85597.05  46242.026250
1     914.258625     629.79    284.468625
2    1077.117600     765.18    311.937600
3    1543.571400    1077.35    466.221400
4   11364.403500    8663.60   2700.803500


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [20]:
# Check your work using the totals above
# Confirm that Order ID 2742071 had a total price of $152,811.89
order_2742071_total = df[df['order_id'] == 2742071]['line_price'].sum()
print(f"Order 2742071 Total: ${order_2742071_total:.2f}")

# Confirm that Order ID 2173913 had a total price of $162,388.71
order_2173913_total = df[df['order_id'] == 2173913]['line_price'].sum()
print(f"Order 2173913 Total: ${order_2173913_total:.2f}")

# Confirm that Order ID 6128929 had a total price of $923,441.25
order_6128929_total = df[df['order_id'] == 6128929]['line_price'].sum()
print(f"Order 6128929 Total: ${order_6128929_total:.2f}")

Order 2742071 Total: $152811.90
Order 2173913 Total: $162388.72
Order 6128929 Total: $923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [23]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
client_summary = df[df['client_id'].isin(top_client_ids)].groupby('client_id').agg(
    total_units=('qty', 'sum'),
    total_shipping=('shipping_price', 'sum'),
    total_revenue=('line_price', 'sum'),
    total_profit=('line_profit', 'sum')
)

# Display the summary DataFrame
print("Client Summary (before formatting):")
print(client_summary)


Client Summary (before formatting):
           total_units  total_shipping  total_revenue  total_profit
client_id                                                          
24741           239862      5126448.37   8.226889e+07  3.657999e+07
33615            64313      1828984.89   8.377309e+06  2.201995e+06
38378            73667      3429455.40   1.290655e+07  3.271830e+06
46820            75768      1601448.84   9.743794e+06  2.736603e+06
66037            43018      1395151.85   1.025951e+07  3.255032e+06


In [24]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
def currency_format_millions(x):
    return f"${x/1e6:.2f}M"


In [25]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = ['total_shipping', 'total_revenue', 'total_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    return f"${x/1e6:.2f}M"


# Apply the currency_format_millions function to only the money columns. 
for col in money_columns:
    client_summary[col] = client_summary[col].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
client_summary.columns = ['Total Units', 'Shipping (Millions)', 'Total Revenue (Millions)', 'Total Profit (Millions)']

# Display the formatted DataFrame
print("\nFormatted Client Summary:")
print(client_summary)




Formatted Client Summary:
           Total Units Shipping (Millions) Total Revenue (Millions)  \
client_id                                                             
24741           239862              $5.13M                  $82.27M   
33615            64313              $1.83M                   $8.38M   
38378            73667              $3.43M                  $12.91M   
46820            75768              $1.60M                   $9.74M   
66037            43018              $1.40M                  $10.26M   

          Total Profit (Millions)  
client_id                          
24741                     $36.58M  
33615                      $2.20M  
38378                      $3.27M  
46820                      $2.74M  
66037                      $3.26M  


In [26]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
client_summary_sorted = client_summary.sort_values(by='Total Profit (Millions)', ascending=False)

# Display the sorted DataFrame
print("\nSorted Client Summary by Total Profit:")
print(client_summary_sorted)


Sorted Client Summary by Total Profit:
           Total Units Shipping (Millions) Total Revenue (Millions)  \
client_id                                                             
24741           239862              $5.13M                  $82.27M   
38378            73667              $3.43M                  $12.91M   
66037            43018              $1.40M                  $10.26M   
46820            75768              $1.60M                   $9.74M   
33615            64313              $1.83M                   $8.38M   

          Total Profit (Millions)  
client_id                          
24741                     $36.58M  
38378                      $3.27M  
66037                      $3.26M  
46820                      $2.74M  
33615                      $2.20M  
